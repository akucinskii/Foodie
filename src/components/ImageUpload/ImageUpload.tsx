import Img from "next/image";
import React from "react";

type Props = {
  setImage: (value: string) => void;
  image?: string;
};

async function isBase64UrlImage(base64String: string) {
  const image = new Image();
  image.src = base64String;
  return await new Promise((resolve) => {
    image.onload = function () {
      if (image.height === 0 || image.width === 0) {
        resolve(false);
        return;
      }
      resolve(true);
    };
    image.onerror = () => {
      resolve(false);
    };
  });
}

const ImageUpload = ({ setImage, image }: Props) => {
  const sendFile = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files && e.target.files[0]) {
      // get image from upload input
      const file = e.target.files[0];
      // convert image to base64
      const reader = new FileReader();
      reader.readAsDataURL(file);
      reader.onloadend = () => {
        const base64 = reader.result;
        if (base64) {
          isBase64UrlImage(base64.toString()).then((isImage) => {
            if (isImage) {
              setImage(base64.toString());
            } else {
              alert("Please upload an image");
            }
          });
        }
      };
    }
  };

  return (
    <div className="flex flex-col p-2">
      <div className={`relative ${image && "aspect-square"} h-full w-full`}>
        {image && (
          <Img src={image} alt="uploaded image" className="object-cover" fill />
        )}
      </div>
      <input type="file" accept="image/*" onChange={sendFile} />
    </div>
  );
};

export default ImageUpload;
