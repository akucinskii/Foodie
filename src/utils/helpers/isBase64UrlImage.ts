export async function isBase64UrlImage(base64String: string) {
  const image = new Image();
  image.src = base64String;
  return await new Promise((resolve) => {
    image.onload = function () {
      if (image.height === 0 || image.width === 0) {
        resolve(false);
        return;
      }
      resolve(true);
    };
    image.onerror = () => {
      resolve(false);
    };
  });
}
