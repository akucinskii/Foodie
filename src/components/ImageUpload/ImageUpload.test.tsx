import { fireEvent, render, waitFor } from "@testing-library/react";
import { isBase64UrlImage } from "src/utils/helpers/isBase64UrlImage";
import { beforeEach, describe, expect, Mock, test, vi } from "vitest";
import ImageUpload from "./ImageUpload";

export const testImage =
  "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIAQMAAAD+wSzIAAAABlBMVEX///+/v7+jQ3Y5AAAADklEQVQI12P4AIX8EAgALgAD/aNpbtEAAAAASUVORK5CYII=";

vi.mock("src/utils/helpers/isBase64UrlImage");

beforeEach(() => {
  vi.resetAllMocks();
});

describe("ImageUpload", () => {
  const mockProps = {
    setImage: vi.fn(),
    image: testImage,
  };

  const { container, rerender, getByTestId } = render(
    <ImageUpload {...mockProps} />
  );

  test("should match snapshot", () => {
    expect(container).toMatchSnapshot();
  });

  test("should match snapshot when image is set", () => {
    rerender(<ImageUpload {...mockProps} />);
    expect(container).toMatchSnapshot("image");
  });

  test("should call setImage when image is set", async () => {
    (isBase64UrlImage as Mock).mockResolvedValueOnce(true);

    rerender(<ImageUpload {...mockProps} />);

    const file = new File([testImage], "test.png", { type: "image/png" });
    const input = getByTestId("image-upload-input");

    fireEvent.change(input, {
      target: {
        files: [file],
      },
    });

    await waitFor(() => expect(mockProps.setImage).toBeCalled());
  });

  test("should not call setImage when image is not set", async () => {
    (isBase64UrlImage as Mock).mockResolvedValueOnce(false);
    global.alert = vi.fn();

    rerender(<ImageUpload {...mockProps} />);

    const file = new File([testImage], "test.png", { type: "image/png" });
    const input = getByTestId("image-upload-input");

    fireEvent.change(input, {
      target: {
        files: [file],
      },
    });

    await waitFor(() => expect(mockProps.setImage).not.toBeCalled());
    await waitFor(() => expect(global.alert).toBeCalled());
  });
});
