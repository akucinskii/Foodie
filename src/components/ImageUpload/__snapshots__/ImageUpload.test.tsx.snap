// Vitest Snapshot v1

exports[`ImageUpload > should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col p-2"
  >
    <div
      class="relative aspect-square h-full w-full"
    >
      <img
        alt="uploaded image"
        class="object-cover"
        data-nimg="fill"
        decoding="async"
        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIAQMAAAD+wSzIAAAABlBMVEX///+/v7+jQ3Y5AAAADklEQVQI12P4AIX8EAgALgAD/aNpbtEAAAAASUVORK5CYII="
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <input
      accept="image/*"
      data-testid="image-upload-input"
      type="file"
    />
  </div>
</div>
`;

exports[`ImageUpload > should match snapshot when image is set > image 1`] = `
<div>
  <div
    class="flex flex-col p-2"
  >
    <div
      class="relative aspect-square h-full w-full"
    >
      <img
        alt="uploaded image"
        class="object-cover"
        data-nimg="fill"
        decoding="async"
        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIAQMAAAD+wSzIAAAABlBMVEX///+/v7+jQ3Y5AAAADklEQVQI12P4AIX8EAgALgAD/aNpbtEAAAAASUVORK5CYII="
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <input
      accept="image/*"
      data-testid="image-upload-input"
      type="file"
    />
  </div>
</div>
`;
